portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java

```diff
@@ -0,0 +1,169 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+/** Models an implementation of the TimeRangeGroup model using ArrayList. */
+public class ArrayListTimeRangeGroup implements TimeRangeGroup {
+
+  // This list of all the time ranges will not be sorted within the methods
+  // of this class. The caller of this class, however, can sort the list of all
+  // time ranges after they obtain it from the getter method.
+  public List<TimeRange> allTimeRanges;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public ArrayListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new ArrayList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
```

Comment:
I think there is an issue with this approach. Consider the case where allTimeRanges look like:
1. 1pm-2pm
2. 5pm-6pm
3. 3pm-4pm

And suppose timeRange is 1:30pm-3:30pm.

The desired end state of the internal list is:
1. 1pm-4pm
2. 5pm-6pm

But I think instead this approach will yield:
1. 1pm-3:30pm
2. 5pm-6pm
3. 3pm-4pm

I strongly recommend that you keep the internal list sorted. I think that should allow you to implement the algorithm with fewer boundary conditions and complex cases.

Reply:
I see. I added some sorting logic that sorts existing time ranges by start time ascending.